crates/engine/tree/src/tree/root.rs

```diff
@@ -118,31 +218,389 @@ where
+                }
+            }
+        }
     }
 }
 
-#[allow(dead_code)]
-impl<Factory> StateRootTask<Factory>
+/// Calculate state root from proofs.
+pub fn calculate_state_root_from_proofs<Factory>(
+    view: ConsistentDbView<Factory>,
+    input_nodes_sorted: &TrieUpdatesSorted,
+    input_state_sorted: &HashedPostStateSorted,
+    multiproof: MultiProof,
+    state: HashedPostState,
+) -> ProviderResult<(B256, TrieUpdates, Duration)>
 where
-    Factory: Send + 'static,
+    Factory: DatabaseProviderFactory<Provider: BlockReader> + Clone,
 {
-    fn run(self) -> StateRootResult {
-        while let Ok(state) = self.state_stream.recv() {
-            Self::on_state_update(&self.config.consistent_view, &self.config.input, state);
-        }
+    let started_at = Instant::now();
 
-        // TODO:
-        //    * keep track of proof calculation
-        //    * keep track of intermediate root computation
-        //    * return final state root result
-        Ok((B256::default(), TrieUpdates::default()))
-    }
+    let proof_targets: HashMap<B256, HashSet<B256>> = state
+        .accounts
+        .keys()
+        .map(|hashed_address| (*hashed_address, HashSet::default()))
+        .chain(state.storages.iter().map(|(hashed_address, storage)| {
+            (*hashed_address, storage.storage.keys().copied().collect())
+        }))
+        .collect();
+
+    let account_trie_nodes = proof_targets
+        .into_par_iter()
+        .map_init(
+            || view.provider_ro().unwrap(),
+            |provider_ro, (hashed_address, hashed_slots)| {
+                // Gather and record storage trie nodes for this account.
+                let mut storage_trie_nodes = BTreeMap::default();
+                let storage = state.storages.get(&hashed_address);
+                for hashed_slot in hashed_slots {
+                    let slot_key = Nibbles::unpack(hashed_slot);
+                    let slot_value = storage
+                        .and_then(|s| s.storage.get(&hashed_slot))
+                        .filter(|v| !v.is_zero())
+                        .map(|v| alloy_rlp::encode_fixed_size(v).to_vec());
+                    let proof = multiproof
+                        .storages
+                        .get(&hashed_address)
+                        .map(|proof| {
+                            proof
+                                .subtree
+                                .iter()
+                                .filter(|e| slot_key.starts_with(e.0))
+                                .collect::<Vec<_>>()
+                        })
+                        .unwrap_or_default();
+                    storage_trie_nodes.extend(target_nodes(
+                        slot_key.clone(),
+                        slot_value,
+                        None,
+                        proof,
+                    )?);
+                }
+
+                let storage_root = next_root_from_proofs(storage_trie_nodes, |key: Nibbles| {
```

Comment:
`next_root_from_proofs` function served its purpose, but is not always correct (see https://github.com/paradigmxyz/reth/issues/12859). let's integrate sparse trie and debug it instead